File: packages/foldingosctl/src/registry_import.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"time"
)

var (
	registryCopyBootDiskImage      = copyBootDiskImage
	registryResolveBootDisk        = resolveBootDisk
	registryExpectedImageSizeBytes = releaseImageSizeBytes
	registryReportCopyProgress     = reportRegistryCopyProgress
	registryCopyProgressInterval   = int64(256 * 1024 * 1024)
	registryCopyChunkSize          = 4 * 1024 * 1024
)

func registryImportBootstrap() error {
	if err := requireSupervisorRole(); err != nil {
		return err
	}

	version, err := installedFoldingOSVersionReader()
	if err != nil {
		return err
	}

	if existing, err := loadRegistryEntry(version); err == nil {
		if err := verifyRegistryImageFile(existing.LocalImagePath, existing.ImageSHA256, existing.ImageSizeBytes); err != nil {
			return fmt.Errorf("existing registry image for %s is invalid: %w", version, err)
		}
		fmt.Printf("Registry already contains verified image for FoldingOS %s.\n", version)
		return nil
	} else if !os.IsNotExist(err) {
		return err
	}

	disk, err := registryResolveBootDisk()
	if err != nil {
		return err
	}

	imagePath := registryImagePath(version)
	if err := os.MkdirAll(filepath.Dir(imagePath), 0755); err != nil {
		return err
	}

	reportRegistryImportStarted(version, disk)
	digest, size, err := registryCopyBootDiskImage(disk, imagePath, registryExpectedImageSizeBytes)
	if err != nil {
		reportRegistryImportFailure(version, err)
		return err
	}
	if size != registryExpectedImageSizeBytes {
		return fmt.Errorf("imported image size %d does not match release image size %d", size, registryExpectedImageSizeBytes)
	}

	entry := registryEntry{
		SchemaVersion:      1,
		FoldingOSVersion:   version,
		GitRevision:        embeddedBuildRevisionReader(),
		ImageSHA256:        digest,
		ImageSizeBytes:     size,
		VerificationMethod: "sha256",
		ImportTimestamp:    time.Now().UTC().Format(time.RFC3339),
		RolloutState:       "ready",
		LocalImagePath:     imagePath,
	}
	if err := saveRegistryEntry(entry); err != nil {
		return err
	}
	reportRegistryImportComplete(version, digest)
	fmt.Printf("Imported FoldingOS %s into the supervisor registry.\n", version)
	fmt.Printf("Image SHA-256: %s\n", digest)
	return nil
}

func reportRegistryImportStarted(version, disk string) {
	message := fmt.Sprintf("Registry: copying FoldingOS %s from %s", version, disk)
	emitRegistryStatus(message)
}

func reportRegistryImportComplete(version, digest string) {
	message := fmt.Sprintf("Registry: imported FoldingOS %s (%s)", version, digest)
	emitRegistryStatus(message)
}

func reportRegistryImportFailure(version string, err error) {
	message := fmt.Sprintf("Registry: failed to import FoldingOS %s (%v)", version, err)
	emitRegistryStatus(message)
}

func reportRegistryCopyProgress(written, total int64) {
	emitRegistryStatus(formatRegistryCopyProgress(written, total))
}

func formatRegistryCopyProgress(written, total int64) string {
	percent := int((written * 100) / total)
	return fmt.Sprintf(
		"Registry: copying release image %d MiB / %d MiB (%d%%)",
		written/(1024*1024),
		total/(1024*1024),
		percent,
	)
}

func emitRegistryStatus(message string) {
	fmt.Println(message)
	_ = writeConsole(message + "\n")
}

func copyBootDiskImage(disk, destination string, size int64) (string, int64, error) {
	source, err := os.Open(disk)
	if err != nil {
		return "", 0, err
	}
	defer source.Close()

	temp, err := os.CreateTemp(filepath.Dir(destination), ".registry-image.tmp-")
	if err != nil {
		return "", 0, err
	}
	tempPath := temp.Name()
	cleanup := true
	defer func() {
		if cleanup {
			os.Remove(tempPath)
		}
	}()

	hasher := sha256.New()
	buffer := make([]byte, registryCopyChunkSize)
	var written int64
	var lastReport int64
	for written < size {
		remaining := size - written
		chunkSize := int64(len(buffer))
		if remaining < chunkSize {
			chunkSize = remaining
		}
		n, err := io.ReadFull(source, buffer[:chunkSize])
		if err != nil {
			temp.Close()
			return "", written, fmt.Errorf("copy boot disk image: %w", err)
		}
		if _, err := temp.Write(buffer[:n]); err != nil {
			temp.Close()
			return "", written, err
		}
		if _, err := hasher.Write(buffer[:n]); err != nil {
			temp.Close()
			return "", written, err
		}
		written += int64(n)
		if written == size || written-lastReport >= registryCopyProgressInterval {
			registryReportCopyProgress(written, size)
			lastReport = written
		}
	}
	if err := temp.Sync(); err != nil {
		temp.Close()
		return "", 0, err
	}
	if err := temp.Close(); err != nil {
		return "", 0, err
	}
	if err := os.Rename(tempPath, destination); err != nil {
		return "", 0, err
	}
	cleanup = false
	return hex.EncodeToString(hasher.Sum(nil)), written, nil
}

func verifyRegistryImageFile(path, expectedDigest string, expectedSize int64) error {
	info, err := os.Stat(path)
	if err != nil {
		return err
	}
	if info.Size() != expectedSize {
		return fmt.Errorf("image size %d does not match expected %d", info.Size(), expectedSize)
	}
	file, err := os.Open(path)
	if err != nil {
		return err
	}
	defer file.Close()

	hasher := sha256.New()
	if _, err := io.Copy(hasher, file); err != nil {
		return err
	}
	actual := hex.EncodeToString(hasher.Sum(nil))
	if actual != expectedDigest {
		return errors.New("image SHA-256 does not match registry metadata")
	}
	return nil
}

func resolveBootDisk() (string, error) {
	rootSource, err := output("findmnt", "-n", "-o", "SOURCE", "/")
	if err != nil {
		return "", err
	}
	rootSource = strings.TrimSpace(rootSource)
	if !strings.HasPrefix(rootSource, "/dev/") {
		return "", fmt.Errorf("root source is not a block device: %q", rootSource)
	}

	parentName, err := output("lsblk", "-n", "-o", "PKNAME", rootSource)
	if err != nil {
		return "", err
	}
	parentName = strings.TrimSpace(parentName)
	if parentName == "" || strings.Contains(parentName, "/") {
		return "", fmt.Errorf("could not resolve boot disk from %q", rootSource)
	}
	return filepath.Join("/dev", parentName), nil
}

File: packages/foldingosctl/src/registry_import.rs
use std::fs::{self, File};
use std::io::{Read, Write};
use std::path::Path;

use sha2::{Digest, Sha256};

use crate::identity::read_installed_foldingos_version;
use crate::paths::AppliancePaths;
use crate::process::write_console;
use crate::registry_image::{
    current_import_timestamp, read_embedded_build_revision, registry_image_path,
    save_registry_entry, verify_registry_image_file, RegistryEntry, RELEASE_IMAGE_SIZE_BYTES,
};
use crate::role::require_supervisor_role;
use crate::storage::resolve_boot_disk;

const COPY_PROGRESS_INTERVAL: i64 = 256 * 1024 * 1024;
const COPY_CHUNK_SIZE: usize = 4 * 1024 * 1024;

pub fn import_bootstrap(paths: &AppliancePaths) -> Result<(), String> {
    require_supervisor_role(paths)?;

    let version = read_installed_foldingos_version()?;
    match crate::registry_image::load_registry_entry(paths, &version) {
        Ok(existing) => {
            verify_registry_image_file(
                Path::new(&existing.local_image_path),
                &existing.image_sha256,
                existing.image_size_bytes,
            )
            .map_err(|error| {
                format!("existing registry image for {version} is invalid: {error}")
            })?;
            println!("Registry already contains verified image for FoldingOS {version}.");
            return Ok(());
        }
        Err(error) if error.contains("No such file") || error.contains("not found") => {}
        Err(error) => return Err(error),
    }

    let disk = resolve_boot_disk()?;
    let image_path = registry_image_path(paths, &version);
    if let Some(parent) = image_path.parent() {
        fs::create_dir_all(parent)
            .map_err(|error| format!("create registry images dir: {error}"))?;
    }

    report_registry_import_started(&version, &disk);
    let (digest, size) = match copy_boot_disk_image(&disk, &image_path, RELEASE_IMAGE_SIZE_BYTES) {
        Ok(result) => result,
        Err(error) => {
            report_registry_import_failure(&version, &error);
            return Err(error);
        }
    };
    if size != RELEASE_IMAGE_SIZE_BYTES {
        return Err(format!(
            "imported image size {size} does not match release image size {RELEASE_IMAGE_SIZE_BYTES}"
        ));
    }

    let entry = RegistryEntry {
        schema_version: 1,
        foldingos_version: version.clone(),
        git_revision: read_embedded_build_revision(paths),
        image_sha256: digest.clone(),
        image_size_bytes: size,
        retrieval_url: String::new(),
        verification_method: "sha256".into(),
        import_timestamp: current_import_timestamp(),
        rollout_state: "ready".into(),
        local_image_path: image_path.to_string_lossy().into_owned(),
    };
    save_registry_entry(paths, entry)?;
    report_registry_import_complete(&version, &digest);
    println!("Imported FoldingOS {version} into the supervisor registry.");
    println!("Image SHA-256: {digest}");
    Ok(())
}

pub fn copy_boot_disk_image(
    disk: &str,
    destination: &Path,
    size: i64,
) -> Result<(String, i64), String> {
    let mut source = File::open(disk).map_err(|error| error.to_string())?;
    let parent = destination
        .parent()
        .ok_or_else(|| "destination path has no parent".to_string())?;
    let temp_path = tempfile_in_dir(parent, ".registry-image.tmp-")?;
    let mut cleanup = true;
    let result = (|| -> Result<(String, i64), String> {
        let mut temp = File::create(&temp_path).map_err(|error| error.to_string())?;
        let mut hasher = Sha256::new();
        let mut buffer = vec![0_u8; COPY_CHUNK_SIZE];
        let mut written = 0_i64;
        let mut last_report = 0_i64;
        while written < size {
            let remaining = size - written;
            let chunk_size = (buffer.len() as i64).min(remaining) as usize;
            source
                .read_exact(&mut buffer[..chunk_size])
                .map_err(|error| format!("copy boot disk image: {error}"))?;
            temp.write_all(&buffer[..chunk_size])
                .map_err(|error| error.to_string())?;
            hasher.update(&buffer[..chunk_size]);
            written += chunk_size as i64;
            if written == size || written - last_report >= COPY_PROGRESS_INTERVAL {
                report_registry_copy_progress(written, size);
                last_report = written;
            }
        }
        temp.sync_all().map_err(|error| error.to_string())?;
        drop(temp);
        fs::rename(&temp_path, destination).map_err(|error| error.to_string())?;
        cleanup = false;
        Ok((format!("{:x}", hasher.finalize()), written))
    })();
    if cleanup {
        let _ = fs::remove_file(&temp_path);
    }
    result
}

fn tempfile_in_dir(parent: &Path, prefix: &str) -> Result<std::path::PathBuf, String> {
    for attempt in 0..100 {
        let candidate = parent.join(format!("{prefix}{attempt}"));
        if candidate.exists() {
            continue;
        }
        return Ok(candidate);
    }
    Err("could not create temporary registry image file".into())
}

fn report_registry_import_started(version: &str, disk: &str) {
    emit_registry_status(&format!(
        "Registry: copying FoldingOS {version} from {disk}"
    ));
}

fn report_registry_import_complete(version: &str, digest: &str) {
    emit_registry_status(&format!(
        "Registry: imported FoldingOS {version} ({digest})"
    ));
}

fn report_registry_import_failure(version: &str, err: &str) {
    emit_registry_status(&format!(
        "Registry: failed to import FoldingOS {version} ({err})"
    ));
}

fn report_registry_copy_progress(written: i64, total: i64) {
    emit_registry_status(&format_registry_copy_progress(written, total));
}

fn format_registry_copy_progress(written: i64, total: i64) -> String {
    let percent = ((written * 100) / total) as i64;
    format!(
        "Registry: copying release image {} MiB / {} MiB ({}%)",
        written / (1024 * 1024),
        total / (1024 * 1024),
        percent
    )
}

fn emit_registry_status(message: &str) {
    println!("{message}");
    let _ = write_console(&format!("{message}\n"));
}
